### PR TITLE
Update to Go 1.21.6 and Alpine 3.19

### PR DESCRIPTION
**Pull request by mrfelton**
Update to Go to v1.21.6

There was quite a bit required to get this to work, including:

**Updating Go**

Go has been updated to v1.21.6, and as a result we're also now running on the latest Alpine build, which address a number of known security vulnerabilities.

**Updating all dependencies**

In order to hit all of the transient dependencies that have known security issues, most of the direct dependencies had to be updated to latest versions.

**Switching to doing linting using the golangci-lint official prebuilt docker image**

To update to Go 1.21, we also had to update the linter. I spent a lot of time trying to get the existing linting setup to work, however unfortunately I just could not get it to build cleanly as things are. Given that the maintainers recommend against building from source and recognise that it often doesn't work, instead I have switched to using one of their recommended approaches as per https://golangci-lint.run/usage/install/. For some context, see

- https://github.com/vmware-tanzu/vm-operator/pull/70
- https://github.com/ryancurrah/gomodguard/issues/33
- https://github.com/kubernetes-sigs/cluster-api-provider-cloudstack/pull/226
- https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/1862

**Disabled some lint rules**

With the linter updated, a number of new/changed rules started throwing issues. I fixed a couple of them but it was turning into a distraction. So, I have disabled three rules which we're currently not passing. `depguard`, `goconst`, `gosec`. My Go skill are lacking, so fixing these will take too long for me right now. Maybe someone that's more familiar with Go can re-enable these and fix the offending issues.

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -16,7 +16,7 @@ env:
   # go needs absolute directories, using the $HOME variable doesn't work here.
   GOCACHE: /home/runner/work/go/pkg/build
   GOPATH: /home/runner/work/go
-  GO_VERSION: 1.18.4
+  GO_VERSION: 1.19.13
```
Comment (Roasbeef) [inline .github/workflows/main.yml:19]: Any reason to not go further to 1.21.6?

Comment (mrfelton) [reply .github/workflows/main.yml:19]: I wasn't sure if there is a reason you had things pinned on an earlier version. I went with 1.19 to match what was used in lnd. But I can update it all the way to latest versions.

Comment (Roasbeef) [reply .github/workflows/main.yml:19]: No reason, good reason really, updating to the newer version should also give us some free perf improvements.

File: go.mod [hunk 2]
```diff
@@ -1,5 +1,7 @@
 module github.com/lightninglabs/lndmon
 
+go 1.19
```
Comment (guggero) [inline go.mod:3]: micro nit: leave the version field at the bottom of the file and move it accordingly in the `tools/go.mod`? I know this isn't super consistent across all projects but I'd guess the version at the end is more common (and I tend to move it down there whenever I touch a `go.mod` file)...

File: Dockerfile [hunk 3]
```diff
@@ -1,4 +1,8 @@
-FROM golang:1.18-alpine as builder
+# If you change this value, please change it in the following files as well:
+# /Dockerfile
+# /tools/Dockerfile
+# /.github/workflows/main.yml
+FROM golang:1.19.13-alpine3.18 as builder
```
Comment (guggero) [inline Dockerfile:5]: Any specific reason for also pinning the alpine version? Or just because the default for the Go 1.19 image is an older one?
Looks like if we go to Go 1.21.6 we get alpine 3.19 by default, so I'd suggest just using the `golang:1.21.6-alpine` image here.

**Review by guggero — COMMENTED** [review]
Thanks for the updates!
The package updates you mentioned could go into this PR as well if you feel like it. Or a separate one, whatever you prefer.

**Comment by mrfelton** [comment]
I've updated the PR to move to latest Go version (1.21.6). In order to do that, a lot of the other dependencies needed to be updated, so the scope of the PR has grown somewhat. But this does now address almost all of the known security issues from our dependencies. But there are some caveats - see updated issue description for details.

File: go.mod [hunk 4]
```diff
@@ -154,4 +155,6 @@ require (
 	sigs.k8s.io/yaml v1.2.0 // indirect
 )
 
-go 1.19
+replace google.golang.org/protobuf => github.com/lightninglabs/protobuf-go-hex-display v1.30.0-hex-display
```
Comment (guggero) [inline go.mod:158]: Can you please add this comment above that we also have in other projects, so we know why we need this:

```
// We want to format raw bytes as hex instead of base64. The forked version
// allows us to specify that as an option.
```

File: Makefile [hunk 5]
```diff
@@ -17,6 +17,7 @@ RM := rm -f
 CP := cp
 MAKE := make
 DOCKER_TOOLS = docker run -v $$(pwd):/build lndmon-tools
+DOCKER_LINT = docker run -t --rm -v $$(pwd):/app -w /app golangci/golangci-lint:v1.55.2 golangci-lint run -v
```
Comment (guggero) [inline Makefile:20]: Sorry you lost time on this... The `-tags=tools` in the `tools/Dockerfile` (Line 18) was an oversight on my part in the first place, we removed that in most other projects since it caused this exact problem...
Can you please just remove that too and revert the changes in the `Makefile`?

File: .golangci.yml [hunk 6]
```diff
@@ -22,21 +22,21 @@ linters:
     - bodyclose
     - deadcode
     - decorder
-    - depguard
+    # - depguard
```
Comment (guggero) [inline .golangci.yml:25]: I'm fine with removing `depguard` and `goconst`, they don't seem to be particularly useful to our use case. Can you add a `TODO` to the `gosec` one please? I'll create a PR to fix that today.

File: collectors/htlcs_collector.go [hunk 7]
```diff
@@ -202,12 +202,18 @@ func (h *htlcMonitor) processHtlcEvent(event *routerrpc.HtlcEvent) error {
...
+
+// processHtlcEvent processes all the htlc events we consume from our stream.
+func (h *htlcMonitor) processHtlcEvent(event *routerrpc.HtlcEvent) error {
 	switch e := event.Event.(type) {
 	// If we have received a forwarding event, we add it to our map if it
 	// is not already present. We are ok with duplicate events, because
 	// htlcs are sometimes replayed by the switch, but we want to keep our
 	// earliest timestamp for stats.
 	case *routerrpc.HtlcEvent_ForwardEvent:
+		key, ts := getKeyAndTimestamp(event)
```
Comment (guggero) [inline collectors/htlcs_collector.go:216]: nit: Looks like we could just call this once, at the beginning of the method?

**Review by guggero — COMMENTED** [review]
Thanks a lot for your investigations! Unfortunately the linter is a bit of a pain to handle sometimes...

File: collectors/htlcs_collector.go (hunk 7 above)
Comment (mrfelton) [reply collectors/htlcs_collector.go:216]: The linter complains about that. It classifies the vars key and tx as unused because they aren't used in every single branch of the switch statement. That is why I extracted to a method and then call the method only from the branches where it's used.

Comment (guggero) [reply collectors/htlcs_collector.go:216]: Oh, I see. I thought maybe the linter complained about the length of the function. But then it makes sense to leave it the way it is now.

File: .golangci.yml (hunk 6 above)
Comment (mrfelton) [reply .golangci.yml:25]: Updated

File: Makefile (hunk 5 above)
Comment (mrfelton) [reply Makefile]: Awesome, that does the trick! Updated.

File: go.mod (hunk 4 above)
Comment (mrfelton) [reply go.mod:160]: Added.

**Review by mrfelton — COMMENTED** [review]
All updates made @guggero . Thanks for the pointer on the linter issues. Removing the tools tag worked

File: .golangci.yml (hunk 6 above)
Comment (guggero) [reply .golangci.yml:25]: Can you please remove the two lines instead of commenting them out? The ones without the TODO that is.

Comment (mrfelton) [reply .golangci.yml]: Updated.

**Review by guggero — APPROVED** [review]
Thanks a lot, LGTM :tada:
